[1단계 - 로또 구현] 슬링키(김민기) 미션 제출합니다.

supernovaMK opened the pull request:
안녕하세요 슬링키 입니다!

페어분 레포에서 PR을 보냈어서 다시 PR 보냅니다!

[프로로그 링크](https://prolog.techcourse.co.kr/studylogs/3844)

리뷰어님 반갑습니다.

부족한 부분이 많지만 잘 부탁드립니다!

🏗 아쉬웠던 점

final로 설정할 수 있는 값들을 리팩토링하지 못함
불변성을 보장할 수 있는 값들을 final로 선언하지 못해 아쉬움이 남음.
추후 코드 수정 및 유지보수 시 예기치 않은 변경을 방지하기 위해 final 적용 필요.

예외 메시지 관리 미흡
예외 메시지를 개별적으로 선언하거나 별도의 상수로 관리하지 못함.
일관된 메시지를 유지하기 위해 ErrorMessage 등의 클래스로 관리하는 방안을 고려.

불변 리스트와 sort() 문제
컬렉션을 불변 리스트로 반환하는 방식을 활용했으나, sort()를 호출하는 과정에서 문제가 발생.
List.copyOf() 또는 �stream().toList()를 사용하면 정렬이 불가능하므로, 정렬이 필요한 경우에는 new ArrayList<>(list)와 같은 변환 과정 필요.
이 문제로 인해 불필요한 시간 소모가 발생.

테스트 리팩토링 부족
테스트 코드의 가독성이 아쉬움. (given-when-then 이 없음)

문제 요구사항을 지키지 못한 부분들이 있음
15줄 함수를 맞추려고 하다보니 과하게 코드가 복잡해지는 경향이 아쉬움

m-a-king committed:
docs(README): 기능 요구사항, 구현해야하는 기능 추가

m-a-king committed:
test(Price): 금액 관련 클래스 생성자, 게터 테스트 추가

m-a-king committed:
feat(Price): 금액 관련 클래스 생성

m-a-king committed:
test(Lotto): 로또 번호 관련 테스트 추가

m-a-king committed:
feat(Lotto): 로또 번호 관련 클래스 생성

- 로또 번호 검증 추가

m-a-king committed:
test(LottoMachine): 로또 머신 관련 테스트 추가

- Price 객체를 받는 생성자 로직 테스트

m-a-king committed:
feat(LottoMachine): 로또 머신 클래스 생성

- 로또 금액 상수 추가
- 구입 가능한 로또 개수를 생성자에서 계산

m-a-king committed:
feat(LottoController): 로또 컨트롤러 생성

- 구입금액 입출력
- 전체 로또 티켓 개수 출력
- 전체적인 흐름 관리

m-a-king committed:
docs(README): 구현해야하는 기능 완료 체크

Commit:
docs(README): 구현해야하는 기능 추가

Commit:
feat:난수 생성 기능 추가

Commit:
test(RandomGenerator):난수 생성 테스트 추가

- 난수 값 개수 검증
- 난수 값 범위 검증
- 난수 값 중복 검증

Commit:
feat(Lottos): 로또들을 관리하는 일급 컬렉션 생성

Commit:
test(Lottos): 로또들을 관리하는 기능 테스트 추가

- 로또를 저장하는 기능 검증

Commit:
feat(LottoMachine): 로또를 생성하는 기능 구현

Commit:
feat(Lotto):로또 번호 정렬 기능 & 출력 기능 구현

Commit:
test(Lotto): 로또 관련 테스트 추가

- 로또 숫자 개수 검증
- 로또 숫자 정렬 검증
- 로또 숫자 출력 검증

Commit:
feat(LottoController): 로또 목록 생성 & 출력

m-a-king committed:
docs(README): 구현해야하는 기능 완료 체크

m-a-king committed:
feat(Lotto): 로또 번호 중복 검증 추가

m-a-king committed:
refactor(LottoMachine): 불필요한 코드 삭제

m-a-king committed:
refactor(RandomGenerator): 난수 생성 로직 변경

- 강력한 SecureRandom

m-a-king committed:
feat(LottoRanking): 로또 순위 ENUM 구현

- 1 ~ 5 순위와 꽝으로 구분
- 맞힌 개수와 보너스 번호를 기준으로 순위 반환 정적 메서드 구현

m-a-king committed:
test(LottoRanking): 로또 순위 테스트 추가

- 나올 수 있는 모든 경우를 테스트

m-a-king committed:
feat(WinningNumberWithBonusNumber): 당첨번호와 보너스번호를 래핑하는 클래스 구현

- 중복 검증 구현

m-a-king committed:
feat(LottoResult): 로또 결과 클래스 구현

- 순위를 키로 하는 Map을 인스턴스 변수로 갖는다.
- 초기화 정적 메서드 구현

m-a-king committed:
feat(LottoWinningChecker): 로또 결과 계산 클래스 구현

m-a-king committed:
feat(LottoController): 당첨 번호와 보너스볼 입력, 결과 계산 호출 추가

m-a-king committed:
docs(README): 구현 기능에 당첨 통계 계산 추가, 완료한 구현 기능 체크

Commit:
refactor(Lotto): 로또 숫자 검증 로직 오류 수정

Commit:
refactor(LottoMachine): 로또머신 메서드명 수정

Commit:
test(RandomGenerator): 테스트 도구 변경

Commit:
test(LottoMachine): 로또 생성 테스트 추가

Commit:
feat(LottoRanking): 게터 추가

Commit:
test(Lotto): 로또 번호 중복 테스트 추가

Commit:
feat(LottoWinningChecker): 생성자 접근 제한

Commit:
feat(LottoResult): 총 상금 계산 로직 구현

- 맵 구현체 변경(LinkedHashMap의 순서 보장을 사용)

Commit:
test(LottoWinningChecker): 로또 결과 확인 테스트 추가

- 모든 케이스 검증

Commit:
test(LottoResult): 로또 결과 테스트 추가

- 초기화된 객체의 Map 검증
- 결과에 로또 순위 개수 추가 검증

Commit:
feat(LottoController): 로또 당첨 계산 및 출력 로직 구현

- 수익률 포함

m-a-king committed:
docs(README): 구현해야하는 기능 완료 체크

m-a-king committed:
Merge remote-tracking branch 'origin/step1' into step1

m-a-king committed:
refactor: 코드 포맷 정리

m-a-king committed:
refactor: 코드 스타일 개선

m-a-king committed:
test(LottoResult): 총 상금 계산 테스트 추가

m-a-king committed:
test(Lotto): 불변 리스트 문제 해결

- 가변 리스트로 변경

m-a-king committed:
feat(Price): 로또 구입 금액 입력값 검증

- 1000원 단위, 0원 이상

m-a-king committed:
test(Price): 예외 클래스 변경

m-a-king committed:
test(WinningNumberWithBonusNumber): 당첨 번호, 보너스 번호 중복 예외 테스트 추가

Gomding commented on README.md:
```diff
@@ -1,3 +1,73 @@
 # java-lotto
```
> final로 설정할 수 있는 값들을 리팩토링하지 못함
> 불변성을 보장할 수 있는 값들을 final로 선언하지 못해 아쉬움이 남음.
> 추후 코드 수정 및 유지보수 시 예기치 않은 변경을 방지하기 위해 final 적용 필요.

어떤 부분인지 공유해주실 수 있을까요? 😃 
함께 고민해보면 좋을 것 같아요~

supernovaMK replied:
사용자의 입력 값을 받아오거나, 지난 당첨 번호의 값, 보너스 번호의 값 등이 있을 것 같습니다. 생각하다 보니 사용자가 입력한 값을 객체들이 사용할 때 주로 final을 사용하면 좋을 것 같다고 생각이 듭니다.

Gomding replied:
👍 👍 👍 

final 을 선언하지 못했다는것은 시간이 부족해서 였을까요? 😃

supernovaMK replied:
넵 제출 시에 final을 붙히기 시작했는데 페어분과 다음에는 프로그래밍 과정에서 인지를 하고 한다면 더 좋을 것 같다는 이야기가 나와서 회고에 적어두었습니다!

Gomding commented on README.md:
```diff
@@ -1,3 +1,73 @@
 # java-lotto
```
> 예외 메시지 관리 미흡
> 예외 메시지를 개별적으로 선언하거나 별도의 상수로 관리하지 못함.
> 일관된 메시지를 유지하기 위해 ErrorMessage 등의 클래스로 관리하는 방안을 고려.

👍 👍 👍
예외메세지는 다음 요청때 반영해주시는걸까요? 😃 

추가로 `ErrorMessage` 같은 클래스에 에러 메세지를 관리하면 어떤 장단점이 있을까요?

supernovaMK replied:
예외 처리 메시지에 관해서 한 곳에서 일관적으로 관리 할 수  있는 장점이 있을 것 같습니다.

단점으로는 잘 떠오르지 않는 것 같습니다, 혹시 단점인 부분들이 어떤 부분이 있을까요?

supernovaMK replied:
더 생각을 해보았는데, 에러 메시지가 많아진다면, ErrorMessage 클래스가 너무 많은 내용을 가지고 있다? 정도 생각이 듭니다!

Gomding replied:
개인적인 경험으로는 에외 메세지를 재사용하는 경우는 거의 없었던것 같아요 ㅎㅎ..
클라이언트에게 보여주는 응답이 아니라면요

만약 하나의 클래스에서 관리한다면 여러 객체들이 `ErrorMessage(가명)` 객체를 의존하게 될 것 같아요
`ErrorMessage`에서 수정이 일어나면 그것을 의존하고있는 다른 객체에 변화를 줄수도 있겠죠
- A클래스와 B클래스는 ErrorMessage에 있는 동일한 메세지를 의존하고있음
- A개발자가 A클래스에서 에러메세지 변경이 필요해서 ErrorMessage 의 메세지를 수정함
- B개발자가 개발한 B클래스는 갑자기 다른 에러메세지가 노출되고있음 🤔

supernovaMK replied:
오오 이런 생각까지는 해보지 못했던 것 같습니다.

객체에 변화를 주는지를 항상 주의해야겠네요!

Gomding commented on README.md:
```diff
@@ -1,3 +1,73 @@
 # java-lotto
```
> 문제 요구사항을 지키지 못한 부분들이 있음
> 15줄 함수를 맞추려고 하다보니 과하게 코드가 복잡해지는 경향이 아쉬움

코드가 복잡해진다고 느낀 포인트를  공유해주시면 좋겠어요!!

supernovaMK replied:
15줄 안에서 해결 하다보니, 로직이 길어지는 경우에 과하게 메서드 분리를 시도했던 점들이 있는 것 같습니다. 

15줄을 넘기더라도 해당 메서드가 15줄을 넘어야 할 이유가 있다면, 지키지 못하는 상황도 생길 수 있다고 생각합니다만, 한 메서드가 가지고 있는 역할이 너무 많지는 않은지 고민해보는 의식적인 연습이 필요하다고 느꼈습니다.

Gomding replied:
이미 잘 알고 계시는군요 😳
의식적인 연습을 통해 슬링키만의 기준을 만들어보는게 중요하다 생각해요!

Gomding commented on README.md:
```diff
@@ -1,3 +1,73 @@
 
 로또 미션 저장소
+
+기본) 기능 요구사항
+- 로또 구입 금액을 입력하면 구입 금액에 해당하는 로또를 발급해야 한다.
+- 로또 1장의 가격은 1000원이다.
```
마크다운 문법을 활용하면 체크박스를 만들수도 있어요!
`- [ ] 이런저런 기능이다` 

```suggestion
기본) 기능 요구사항
- [ ] 로또 구입 금액을 입력하면 구입 금액에 해당하는 로또를 발급해야 한다.
- [ ] 로또 1장의 가격은 1000원이다.
```

supernovaMK replied:
넵! 알려주셔서 감사합니다!

Gomding commented on src/main/java/Lotto.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    private void validateDistinct(List<Integer> numbers) {
+        if (numbers.stream().distinct().count() == 6) {
```
6이라는 매직넘버를 상수로 표현하여 의미를 부여하는것은 어떻게 생각하시나요?
validateSize 메서드에도 6이 존재하니 재사용도 가능할 것 같네요!

supernovaMK replied:
좋다고 생각합니다! 놓친 부분들이 많았군요!

Gomding replied:
요것도 반영이 안된것 같아요 😭

supernovaMK replied:
반영해서 커밋해두었습니다! 

상수화하는 과정에 익숙해져보겠습니다!

Gomding commented on src/main/java/Lotto.java:
```diff
@@ -0,0 +1,42 @@
+        throw new IllegalArgumentException();
+    }
+
+    @Override
+    public String toString() {
+        return "[" + String.join(", ", numbers.stream().map(String::valueOf).toList()) + "]";
+    }
```
toString 메서드를 재정의한 이유가 궁금해요~
만약 UI에서 노출하기 위한것이 목적이라면 `toString` 대신 다른 방법을 찾아보면 좋을것 같아요!

`toString` 은 대다수의 개발자가 객체 내부를 상태를 알고싶을 때 디버깅용도로 사용해요
슬링키가 사용한 목적과 대다수가 인지하는것이 다르다면 언젠가 혼란이 발생할 수 있어요
이펙티브 자바 책이 강의장에 있다면 아이템12(toString을 항상 재정의하라)를 읽어보는것도 좋습니다!

supernovaMK replied:
toString의 목적을 다시 한번 생각해보겠습니다!

Gomding replied:
이 코멘트에 대한 슬링키의 의견 공유해주시면 좋겠어요! 🙏

supernovaMK replied:
다음 설계에서는 OutputView에서 다룰 수 있게 설계하여, toString은 사용하지 않게끔 하려고 합니다.

toString이 간단한 값을 출력하고 싶을때 사용한다고 생각했는데, 고민이 많이 필요했던 부분이었던 것 같습니다! 

감사합니다!

Gomding replied:
현재 PR에서 반영해주셔도 좋을것 같아요!

supernovaMK replied:
넵! toString을 사용하지 않고 로직을 구현하여 커밋하였습니다!

Gomding commented on src/main/java/LottoController.java:
```diff
@@ -0,0 +1,132 @@
+
+public class LottoController {
+
+    public void start() throws IOException {
+        BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(System.in));
+        
+        Price price = getPrice(bufferedReader);
+
+        LottoMachine lottoMachine = buyLottos(price);
+        Lottos lottos = printLottos(lottoMachine);
+
+        LottoResult lottoResult = calculateLottoResult(bufferedReader, lottos);
+
+        printLottoResults(lottoResult);
+        printRateOfReturn(lottoResult, price);
+    }
```
메서드 추출을 잘 해주셨네요 👍 
start() 메서드 가독성이 좋다고 생각했어요~
각 행위별로 메서드를 잘 표현해주셨어요 😃

Gomding commented on src/main/java/LottoMachine.java:
```diff
@@ -0,0 +1,29 @@
+
+    private final int ticket;
+
+    private static final int LOTTO_PRICE = 1000;
+    private static final int LOTTO_NUMBER_START = 1;
+    private static final int LOTTO_NUMBER_END = 45;
+    private static final int LOTTO_NUMBER_COUNT = 6;
```
상수로 매직넘버를 잘 표현해주셨네요!! 👍

Gomding commented on src/test/java/LottoMachineTest.java:
```diff
@@ -0,0 +1,21 @@
+        LottoMachine lottoMachine = new LottoMachine(new Price("10000"));
+        assertThat(lottoMachine.generateLottos().getLottos().size()).isEqualTo(lottoMachine.getTicket());
+    }
+}
```
파일 끝 개행이 없을 때 github에서 아래와 같이 표시되는데요
[이 글](https://hyeon9mak.github.io/github-no-newline-at-a-end-of-file/)을 읽고 해결해보면 좋겠어요!

![image](https://github.com/user-attachments/assets/c85b1f2c-b190-483b-ac2f-208532966cb5)

supernovaMK replied:
오 감사합니다! End Of Line 에 대해서 새롭게 배워가네요! 인텔리제이에서 자동으로 설정해주는 부분 체크 했습니다!

supernovaMK replied:
Pending 이슈 테스트

Gomding commented on src/test/java/LottoMachineTest.java:
```diff
@@ -0,0 +1,21 @@
+import org.junit.jupiter.api.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+class LottoMachineTest {
```
아래와 같이 클래스에도 `@DisplayName` 을 사용할 수 있어요~
```suggestion
@DisplayName("LottoMachine 테스트")
class LottoMachineTest {
```

supernovaMK replied:
클래스에서도 @DisplayName을 설정할 수 있군요! 처음 알게 된 사실이네요. 잘 활용해보겠습니다!

Gomding commented on src/test/java/PriceTest.java:
```diff
@@ -0,0 +1,32 @@
+                .isInstanceOf(IllegalArgumentException.class);
+    }
+
+    @Test
+    @DisplayName("Price getter")
+    void getterTest() {
```
getter 메서드에 대한 테스트가 필요하다고 느끼신 이유가 궁금해요!

supernovaMK replied:
문제 요구 사항에서 모든 로직에 관한 테스트를 작성하여라. 라는 부분에서 페어분과 처음에 고민을 나눈 부분입니다. getter로 값을 가져오는 메서드 역시 , 로직일까? 라는 고민을 했던 것 같습니다. 

페어분과 계속 생각하다 보니 getter는 로직이 아니다라고 생각이 들어 이후에는 테스트 하지 않았습니다.

찰리는 어떻게 생각하시나요?

Gomding replied:
> 이 코멘트에 대한 슬링키의 의견 공유해주시면 좋겠어요! 🙏



여기두요! ㅎㅎ

Gomding replied:
Pending 상태라 안보였군요 ㅎㅎ

저는 테스트 코드의 목적을 고민해볼 것 같네요!
테스트 코드를 작성하는 목적은 무엇일까요?

모든 로직에 관한 테스트를 작성하는 것은 맞습니다!
학습을 위해 규칙을 강하게 적용하면서 슬링키 만의  테스트 코드 작성시 목적과 장단점, 작성하는 기준을 세우는것이 중요하다고 생각해요~!

supernovaMK replied:
넵! 테스트에서도의 규칙을 정하는 과정이 필요해보이네요!

저만의 테스트 기준을 다양한 목적의 테스트를 통해 단단하게 만들어보겠습니다 :)

Gomding replied:
덧붙이자면 저는 테스트 코드를 작성하는 목적은 안정감을 느끼기 위함이에요
작성한 코드가 비즈니스 규칙에 맞게 동작하는지 보장받기 위해서요
getter 같은 경우는 테스트를 작성안해도 불안함을 느끼지 않는다고 생각했어요 😃

Gomding commented on src/test/java/RandomGeneratorTest.java:
```diff
@@ -0,0 +1,33 @@
+        }
+    }
+
+    @Test
+    @DisplayName("값들의 중복 테스트")
```
테스트 이름을 더 풀어서 적어보면 좋겠어요~
어떤상황에 무엇을 검증하는지요 ㅎㅎ

```suggestion
    @DisplayName("생성하는 난수들은 중복없이 고유한 값으로 생성된다")
```

supernovaMK replied:
넵 확인했습니다!

Gomding commented on src/test/java/RandomGeneratorTest.java:
```diff
@@ -0,0 +1,33 @@
+class RandomGeneratorTest {
+    @Test
+    @DisplayName("난수 값 6개 생성 테스트")
+    void 난수_값_6개_생성_테스트() {
```
메서드 이름을 한글로 적는다면 `@DisplayNameGeneration` 애너테이션을 활용해볼 수 있겠어요~
(그럼 `@DisplayName` 대신 사용이 가능할것 같네요~)

supernovaMK replied:
넵! 다음 테스트 작성시 활용해보겠습니다!

Gomding reviewed (changes requested):
브랜치 변경 후 재요청입니다!

> 안녕하세요 슬링키!
첫 미션의 함께하게된 리뷰어 찰리입니다 🙌
미션 꼼꼼하게 잘 구현해주셨네요~ 
고생하셨어요! 💪 
> 
> 질문에 대한 답변과 함께 몇가지 코멘트 남겼으니 확인 부탁드려요 😃
궁금한 점 있으면 언제든 DM 이나 코멘트 남겨주세요 🙏

Commit:
refactor: 전체적인 코드 스타일 리팩토링

Commit:
refactor: 전체적인 코드 스타일 리팩토링

Gomding commented on src/main/java/Price.java:
```diff
@@ -19,14 +19,14 @@ private void validateLottoPriceUnit(String value) {
             return;
```
리팩토링 시에도 커밋 제목은 자세히 남겨주시는것이 좋아요! ㅎㅎ
협업하는 사람에게 많은 정보가 됩니다~!

`전체적인 코드 스타일 리팩터링` 은 어떤 작업을 했는지 코드를 전부 확인해봐야합니다 😭 
(먼 미래에 작업의 히스토리 추적이 필요할 때도 힘들구요 😞 )
페어 프로그래밍할 때와 동일하게 작은 단위로 커밋을 남기는 습관을 유지하는것이 좋아요!!

supernovaMK replied:
넵!!! 좋은 피드백 감사합니다!

Commit:
docs(README): 구현해야하는 기능 목록 스타일 수정

Gomding commented on src/main/java/Price.java:
```diff
@@ -19,14 +19,14 @@ private void validateLottoPriceUnit(String value) {
             return Integer.parseInt(value);
         } catch (NumberFormatException e) {
-            throw new IllegalArgumentException();
+            throw new IllegalArgumentException("Invalid number: " + value);
```
예외 메세지를 더 친절하게 작성해주시는게 좋아요!
사용자에게 어떤 에러상황인지 알려주고 어떻게 하면 현재 상황을 해결할 수 있는지 알려주면
친절한 예외메세지가 될 수 있습니다 😃 

아래 글을 참고해보시면 좋겠어요!
- https://hyeon9mak.github.io/writing-helpful-error-messages/

supernovaMK replied:
예외 상황에 관해서 자세히 작성하는 연습이 제 자신에게 더 필요해 보이네요, 사용자 입장에서 생각해보도록 노력해야겠네요.

좋은 자료 공유해주셔서 감사합니다!

Gomding commented on src/test/java/RandomGeneratorTest.java:
```diff
@@ -1,21 +1,22 @@
 import static org.assertj.core.api.Assertions.assertThat;
 
-
+@DisplayNameGeneration(DisplayNameGenerator.ReplaceUnderscores.class)
```
해당 애너테이션을 사용하셨다면 전체적으로 `@DisplayName` 은 제거해주셔도 좋을것 같아요~

supernovaMK replied:
넵! 체크해주셔서 감사합니다!

Gomding commented on src/main/java/Lotto.java:
```diff
@@ -0,0 +1,42 @@
+
+public class Lotto {
+
+    private final List<Integer> numbers;
```
Price와 동일하게 로또의 번호도 객체로 포장해볼 수 있곘어요 😄

supernovaMK replied:
충분히 포장해보았다고 생각했었는데, 아직 익숙하지 않았던 것 같습니다.

좋은 피드백 감사합니다!

Gomding commented on src/test/java/LottoWinningCheckerTest.java:
```diff
@@ -0,0 +1,73 @@
+    @ParameterizedTest
+    @DisplayName("로또 결과를 확인할 수 있다")
+    @MethodSource("lottoDataProvider")
+    void 로또_결과를_확인할_수_있다(List<Integer> myNumbers, List<Integer> winningNumbers, int bonusNumber, LottoRanking ranking, int expectedCount) {
+        Lotto lotto = new Lotto(myNumbers);
+        Lotto winningNumbersLotto = new Lotto(winningNumbers);
+
+        Lottos lottos = new Lottos();
+        lottos.addLotto(lotto);
+
+        WinningNumberWithBonusNumber winningNumberWithBonusNumber = new WinningNumberWithBonusNumber(winningNumbersLotto, bonusNumber);
+
+        LottoResult lottoResult = LottoWinningChecker.calculateResult(lottos, winningNumberWithBonusNumber);
+
+        assertThat(lottoResult.result().get(ranking)).isEqualTo(expectedCount);
```
말씀해주신대로 given when then 을 사용해서 가독성을 높일 수 있겠네요!

supernovaMK replied:
넵! 좋은 피드백 감사합니다!

Gomding reviewed (changes requested):
안녕하세요 슬링키!
의견과 함께 몇가지 코멘트 추가로 남겼는데 확인 부탁드려요! 🙏 
궁금한 점 있으면 언제든 DM이나 코멘트 남겨주세요!!
화이팅입니다~

supernovaMK reviewed (commented):
Pending 문제가 있어서 답글이 몇개가 빠졌던 것 같습니다! 다시 올렸습니다!

Commit:
refactor(Lotto): 상수화로 스타일 변경

Commit:
refactor(Lotto): toString 오버라이딩 삭제

Commit:
test(Lotto): 출력 로직 테스트 삭제

- toString 로직이 모두 삭제되었기에, 출력 메서드 테스트 삭제
- @DisplayName 대신 @DisplayNameGenration 사용으로 변경

Commit:
test: 테스트 스타일 변경

- @DisplayName 대신 @DisplayNameGenration 사용으로 변경

Commit:
refactor(Price): 에러 메시지를 더 구체적으로 수정

Commit:
refactor(Numbers): 숫자를 담는 값 포장

- 로또 번호 6자리를 담는 List 형태의 값을 포장하였습니다

Commit:
test: given,when,then 주석 추가

- given,when,then 구조를 사용하여 테스트의 가독성 향상

Commit:
refactor(Lotto): 상수화로 로직 변경

Commit:
refactor(WinningNumberWithBonusNumber): 오류메시지 구체적으로 변경

supernovaMK commented:
기존에 사용하던 로또 번호 값들을 포장하여 리팩토링 하는 과정에서, 수정이 많이 생겼습니다.
테스트 코드 역시, 많은 수정이 생기게 되었던 것 같습니다.

이번 주차가 유닛 테스트 주제가 있었던 만큼, 처음 진행할 때 테스트 코드를 먼저 작성하려는 노력을 많이 하였습니다.
하지만 값 포장 처럼,주요한 타입 혹은 메서드 기능이 변경되었을 때 테스트 코드에 다시 들어가 수정을 하는 일이 빈번했습니다.

결국 코드를 다 작성한 후에 테스트 코드를 작성하게 되는 상황이 생겼던 것 같습니다.

테스트 코드를 어떤 방식으로 구현하면, 이 부분들을 줄일 수 있을까요?

supernovaMK committed:
docs(README): 구현해야하는 기능 목록 스타일 수정

Gomding commented on src/main/java/Lotto.java:
```diff
@@ -4,6 +4,7 @@
 public class Lotto {
 
     private final List<Integer> numbers;
+    private static final int LOTTO_COUNT=6;
```
사소한 부분이지만
공백으로 구분해주면 좋을 것 같아요!

IntteliJ를 사용하고있다면
코드 포맷팅 기능을 사용하면 편리합니다 ㅎㅎ
- https://ddiyeon.tistory.com/87

```suggestion
    private static final int LOTTO_COUNT = 6;
```

Gomding commented on src/main/java/Price.java:
```diff
@@ -19,14 +19,14 @@ private void validateLottoPriceUnit(String value) {
         }
 
-        throw new IllegalArgumentException("Invalid lotto price value: " + value);
+        throw new IllegalArgumentException("로또 단위는 1000원 입니다. 1000원 단위로 입력해주세요 입력된 (" + value + ")는 1000원 단위의 수가 아닙니다.");
```
String의 format 메서드를 사용하면 로또의 가격이 변경됐을 때
수정할 부분이 줄어들 수 있어서 추천드립니다! 😉 
```suggestion
        throw new IllegalArgumentException(String.format("로또 단위는 %d원 입니다. %d원 단위로 입력해주세요 입력된 (%d)는 %d원 단위의 수가 아닙니다.", LOTTO_PRICE, LOTTO_PRICE, value, LOTTO_PRICE));
```

Gomding commented on src/main/java/Numbers.java:
```diff
@@ -0,0 +1,19 @@
+import java.util.Comparator;
+import java.util.List;
+
+public class Numbers {
+    private List<Integer> numbers;
```
앗 포장하는것은 로또번호전체가 아닌 `하나의 단일 숫자` 를 생각하긴 했어요 😃 
단일 로또 번호를 포장하면 숫자에 대한 제약사항을 해당 객체에 넣을 수도 있겠어요~!

```java
public class LottoNumber {
    private final int number;

    public 생성자(final int number) {
        숫자에 대한 검증(number);
        this.number = number
    }
}
```

supernovaMK replied:
아아 그렇군요! 좋은 생각 나눠주셔서 감사합니다!

Gomding commented on src/main/java/Numbers.java:
```diff
@@ -0,0 +1,19 @@
+import java.util.List;
+
+public class Numbers {
+    private List<Integer> numbers;
```
해당 필드는 final 이 아닌 이유가 궁금해요!

supernovaMK replied:
final로 변경하여 사용해도 좋겠군요! 꼼꼼하게 체크해주셔서 감사합니다!

Gomding commented on src/test/java/LottoMachineTest.java:
```diff
@@ -10,13 +10,18 @@ class LottoMachineTest {
     void 로또머신은_로또_구매개수를_갖고있다() {
+        //given
         LottoMachine lottoMachine = new LottoMachine(new Price("10000"));
+        //when & then
```
변경사항이 많았을 텐데 고생하셨습니다!! 👍 👍 👍

Gomding commented on src/main/java/Lotto.java:
```diff
@@ -27,7 +27,7 @@ private void validateDistinct(List<Integer> numbers) {
 
     private static void validateSize(List<Integer> numbers) {
-        if (numbers.size() == 6) {
+        if (numbers.size() == LOTTO_COUNT) {
```
꼼꼼하게 변경해주셨네요~

Gomding commented on src/main/java/WinningNumberWithBonusNumber.java:
```diff
@@ -15,7 +15,7 @@ private void validate(Lotto winningNumber, int bonusNumber) {
     private void validateDistinct(Lotto winningNumber, int bonusNumber) {
         if (winningNumber.getNumbers().contains(bonusNumber)) {
-            throw new IllegalArgumentException("Winning number " + winningNumber + " is already in use");
+            throw new IllegalArgumentException("당첨번호 " + winningNumber + "는 이미 존재하는 번호입니다. 보너스 번호와 당첨번호는 중복될 수 없습니다.");
```
👍 👍 👍 👍 👍

Gomding commented on src/main/java/Price.java:
```diff
@@ -0,0 +1,36 @@
+public class Price {
+
+    public static final int LOTTO_PRICE = 1000;
```
`LOTTO_PRICE = 1000` 이라는 개념이 다른 클래스에도 있는것 같아요!
만약 로또 가격이 변경된다면 여러곳을 변경해야 비로소 로또가격 규칙이 적용되겠죠 😭 
한 곳을 의존하게 하면 변경이 쉬워질 것 같아요!

supernovaMK replied:
오 좋습니다! 리팩토링 시도해볼게요!

Gomding reviewed (approved):
안녕하세요 슬링키!
리뷰 반영하느라 고생하셨어요 👍 👍 
이번 단계는 이만 머지하겠습니다 🙏 
추가로 남긴 코멘트는 다음 단계 진행하면서 함께 반영해주세요~!!
다음 단계도 화이팅입니다 🔥
